vizro-core/src/vizro/models/_utils.py

```diff
@@ -0,0 +1,167 @@
+import inspect
+import subprocess
+import textwrap
+from dataclasses import dataclass
+from typing import Any, Optional, Set
+
+from vizro.managers import model_manager
+
+ACTIONS_CHAIN = "ActionsChain"
+ACTION = "actions"
+
+TO_PYTHON_TEMPLATE = """
+############ Imports ##############
+import vizro.plotly.express as px
+import vizro.tables as vt
+import vizro.models as vm
+import vizro.actions as va
+from vizro import Vizro
+from vizro.models.types import capture
+from vizro.managers import data_manager
+{extra_imports}
+
+{callable_defs}
+{data_settings}
+
+########### Model code ############
+{code}
+"""
+
+CALLABLE_TEMPLATE = """
+####### Function definitions ######
+{callable_defs}
+"""
+
+DATA_TEMPLATE = """
+####### Data Manager Settings #####
+#######!!! UNCOMMENT BELOW !!!#####
+# from vizro.managers import data_manager
+{data_setting}
+"""
+
+
+@dataclass
+class PathReplacement:
+    original: str
+    new: str
+
+
+REPLACEMENT_STRINGS = [
+    PathReplacement("plotly.express", "px."),
+    PathReplacement("vizro.tables", "vt."),
+    PathReplacement("vizro.figures", "vf."),
+    PathReplacement("vizro.actions", "va."),
+    PathReplacement("vizro.charts", "vc."),
+]
+
+
+def _format_and_lint(code_string: str) -> str:
+    # Tracking https://github.com/astral-sh/ruff/issues/659 for proper python API
+    # Good example: https://github.com/astral-sh/ruff/issues/8401#issuecomment-1788806462
+    linted = subprocess.check_output(
+        ["ruff", "check", "--fix", "--exit-zero", "--silent", "--isolated", "-"], input=code_string, encoding="utf-8"
+    )
+    formatted = subprocess.check_output(
+        ["ruff", "format", "--silent", "--isolated", "-"], input=linted, encoding="utf-8"
+    )
+    return formatted
+
+
+def _clean_module_string(module_string: str) -> str:
```

Comment:
My last comment got lost somehow.. :(

So the idea of the function is essentially the following:
- iterate through the list of replacements, if you match a substring, take this as the module replacement
- if not replacement is found, then return ""


The idea is to match substrings, because (as explained in another comment) we do not want to target precise modules for extra stability, ie something like `plotly.express._graph_obj.abcs`.

And yes, there is only ever one match.

Let the code as is for now

Reply:
ok, this makes more sense now 🙂 

Please can you add a brief comment explaining the above?

Personally I would still find it clearer as a Python mapping + iteration rather than list of dataclasses + `next`, like this. But up to you.
```
REPLACEMENT_STRINGS = {"plotly.express": "px.", ...}

def _clean_module_string(module_string: str) -> str:
     # EXPLANATORY COMMENT HERE 
     for module_original, module_alias in REPLACEMENT_STRINGS.items():
        if module_original in module_string:
            return module_alias
```

Reply:
Actually wait, no... I still don't quite understand 😅 Let's have a quick call about it today to get to the bottom of it!

Reply:
Ressovled with simplified data structure for REPLACEMENT_STRINGS